File: Zyntra/Resources/zyntra_api.py
"""
Zyntra Script API — Python Module
Auto-imported into every Python script run by Zyntra.
See the Docs tab in Zyntra for full documentation.
"""

import json, os, sys, atexit
from datetime import datetime

_context_path = os.environ.get("ZYNTRA_CONTEXT", "")
_context = None
_response = {"Notifications": [], "SetClipboard": None, "LaunchGame": []}


def _load_context():
    global _context
    if _context is None and _context_path and os.path.exists(_context_path):
        with open(_context_path, "r", encoding="utf-8") as f:
            _context = json.load(f)


def _flush_response():
    _load_context()
    if _context and _context.get("ResponseFile"):
        with open(_context["ResponseFile"], "w", encoding="utf-8") as f:
            json.dump(_response, f, indent=2)

atexit.register(_flush_response)


# ── Context ──────────────────────────────────────────────────

def get_version() -> str:
    """Returns the current Zyntra version string."""
    _load_context()
    return _context.get("Version", "") if _context else ""

def get_data_dir() -> str:
    """Returns the Zyntra AppData directory path."""
    _load_context()
    return _context.get("DataDir", "") if _context else ""


# ── Accounts ─────────────────────────────────────────────────

def get_accounts() -> list:
    """Returns all Roblox accounts as dicts."""
    _load_context()
    return _context.get("Accounts", []) if _context else []

def get_account(name: str) -> dict | None:
    """Find an account by username or display name."""
    for acc in get_accounts():
        if acc.get("Username") == name or acc.get("DisplayName") == name:
            return acc
    return None

def get_accounts_by_tag(tag: str) -> list:
    """Returns accounts filtered by tag."""
    return [a for a in get_accounts() if a.get("Tag") == tag]

def get_account_count() -> int:
    """Returns the number of accounts."""
    return len(get_accounts())


# ── Apps ─────────────────────────────────────────────────────

def get_apps() -> list:
    """Returns all registered applications as dicts."""
    _load_context()
    return _context.get("Apps", []) if _context else []

def get_app(name: str) -> dict | None:
    """Find an app by name."""
    for app in get_apps():
        if app.get("Name") == name:
            return app
    return None

def get_app_count() -> int:
    """Returns the number of apps."""
    return len(get_apps())


# ── Notifications ────────────────────────────────────────────

def send_notification(title: str, message: str, type: str = "Info"):
    """
    Sends a notification to the Zyntra notification panel.
    type: Info, Success, Warning, Error
    """
    _response["Notifications"].append({
        "Title": title,
        "Message": message,
        "Type": type,
    })


# ── Clipboard ────────────────────────────────────────────────

def set_clipboard(text: str):
    """Sets the Windows clipboard text via Zyntra."""
    _response["SetClipboard"] = text


# ── Game Launch ──────────────────────────────────────────────

def launch_game(account_name: str, place_id: int):
    """
    Launches a Roblox game after script completes.
    account_name: Username or display name of the account to launch with.
    place_id: The Roblox Place ID to join.
    """
    _response["LaunchGame"].append({
        "AccountName": account_name,
        "PlaceId": place_id,
    })


def launch_game_all(place_id: int, tag: str | None = None):
    """
    Launches a game for all accounts, optionally filtered by tag.
    place_id: The Roblox Place ID to join.
    tag: If set, only launch for accounts with this tag.
    """
    accounts = get_accounts_by_tag(tag) if tag else get_accounts()
    for acc in accounts:
        launch_game(acc["Username"], place_id)


# ── Recently Played ─────────────────────────────────────────

def get_recently_played() -> list:
    """Returns list of recently played games as dicts."""
    _load_context()
    return _context.get("RecentGames", []) if _context else []


def get_last_played() -> dict | None:
    """Returns the most recently played game, or None."""
    games = get_recently_played()
    return games[0] if games else None


# ── Utilities ────────────────────────────────────────────────

def log(message: str):
    """Writes a timestamped log line to stdout."""
    ts = datetime.now().strftime("%H:%M:%S")
    print(f"[{ts}] {message}")

File: Zyntra/Resources/zyntra_api.lua
-- ============================================================
--  Zyntra Script API — Lua Module
--  Auto-loaded into every Lua script run by Zyntra.
--  See the Docs tab in Zyntra for full documentation.
-- ============================================================

zyntra = {}

-- ── Context ──────────────────────────────────────────────────

function zyntra.get_version()
    return _zyntra_context.Version or ""
end

function zyntra.get_data_dir()
    return _zyntra_context.DataDir or ""
end

-- ── Accounts ─────────────────────────────────────────────────

function zyntra.get_accounts()
    return _zyntra_context.Accounts or {}
end

function zyntra.get_account(name)
    for _, acc in ipairs(zyntra.get_accounts()) do
        if acc.Username == name or acc.DisplayName == name then
            return acc
        end
    end
    return nil
end

function zyntra.get_accounts_by_tag(tag)
    local result = {}
    for _, acc in ipairs(zyntra.get_accounts()) do
        if acc.Tag == tag then
            table.insert(result, acc)
        end
    end
    return result
end

function zyntra.get_account_count()
    return #zyntra.get_accounts()
end

-- ── Apps ─────────────────────────────────────────────────────

function zyntra.get_apps()
    return _zyntra_context.Apps or {}
end

function zyntra.get_app(name)
    for _, app in ipairs(zyntra.get_apps()) do
        if app.Name == name then
            return app
        end
    end
    return nil
end

function zyntra.get_app_count()
    return #zyntra.get_apps()
end

-- ── Recently Played ─────────────────────────────────────────

function zyntra.get_recently_played()
    return _zyntra_context.RecentGames or {}
end

function zyntra.get_last_played()
    local games = zyntra.get_recently_played()
    if #games > 0 then return games[1] end
    return nil
end

-- ── Notifications ────────────────────────────────────────────

function zyntra.notify(title, message, type)
    type = type or "Info"
    _zyntra_notify(title, message, type)
end

-- ── Clipboard ────────────────────────────────────────────────

function zyntra.set_clipboard(text)
    _zyntra_set_clipboard(text)
end

-- ── Game Launch ──────────────────────────────────────────────

function zyntra.launch_game(account_name, place_id)
    _zyntra_launch_game(account_name, place_id)
end

function zyntra.launch_game_all(place_id, tag)
    local accounts
    if tag then
        accounts = zyntra.get_accounts_by_tag(tag)
    else
        accounts = zyntra.get_accounts()
    end
    for _, acc in ipairs(accounts) do
        zyntra.launch_game(acc.Username, place_id)
    end
end

-- ── Utilities ────────────────────────────────────────────────

function zyntra.log(message)
    _zyntra_log(tostring(message))
end

function zyntra.sleep(ms)
    _zyntra_sleep(ms)
end

-- ── UI ──────────────────────────────────────────────────────

zyntra.ui = {}

function zyntra.ui.create_tab(name, icon)
    icon = icon or ""
    return _zyntra_ui_create_tab(name, icon)
end

function zyntra.ui.add_label(tab, text, opts)
    opts = opts or {}
    _zyntra_ui_add_label(tab, text, opts.font_size or 14, opts.bold or false)
end

function zyntra.ui.add_button(tab, text, callback)
    _zyntra_ui_add_button(tab, text, callback)
end

function zyntra.ui.add_text_input(tab, id, opts)
    opts = opts or {}
    _zyntra_ui_add_text_input(tab, id, opts.placeholder or "")
end

function zyntra.ui.add_separator(tab)
    _zyntra_ui_add_separator(tab)
end

function zyntra.ui.add_progress(tab, value, label)
    label = label or ""
    _zyntra_ui_add_progress(tab, value, label)
end

function zyntra.ui.add_checkbox(tab, id, text, checked)
    if checked == nil then checked = false end
    _zyntra_ui_add_checkbox(tab, id, text, checked)
end

function zyntra.ui.add_dropdown(tab, id, label, options, selected_index)
    selected_index = selected_index or 1
    _zyntra_ui_add_dropdown(tab, id, label, options, selected_index)
end

function zyntra.ui.get_value(tab, id)
    return _zyntra_ui_get_value(tab, id)
end

return zyntra
